### ISSUE TITLE
LabelSelector type-checks with empty component Lists

### ISSUE THREAD

**Issue by ari-becker** [arianvp/dhall-kubernetes#78]
I'm currently trying to fit [external-dns](https://github.com/kubernetes-incubator/external-dns/blob/master/docs/tutorials/aws.md) into our Dhall-ified pipeline, and I replicated the mistake in their tutorial wherein a `DeploymentSpec` is created without a `LabelSelector` (since their `Deployment` doesn't have any labels in its `metadata`).

Currently, a `DeploymentSpec` bundles in its own default `LabelSelector`, which bundles in empty lists for `matchLabels` and `matchExpressions`. Due to `--omitEmpty`, this succeeds at type-checking, and the enveloping `Deployment` also type-checks.

However, attempting to actually apply the `Deployment` results in `error: error validating "STDIN": error validating data: ValidationError(Deployment.spec): missing required field "selector" in io.k8s.api.apps.v1.DeploymentSpec`.

Not sure what a proper solution is considering [dhall-lang#691](https://github.com/dhall-lang/dhall-lang/issues/691) precluded language-level support for non-empty `List`s.

**Comment by f-f** [comment]
@ari-becker I think this is another case in which the Swagger types are "lying" (see #8), and we should integrate that type information by hand here:

https://github.com/dhall-lang/dhall-kubernetes/blob/4ad58156b7fdbbb6da0543d8b314df899feca077/dhall-kubernetes-generator/src/Dhall/Kubernetes/Convert.hs#L42-L47

**Comment by ari-becker** [comment]
Reading the Kubernetes API documentation: 
> A label selector is a label query over a set of resources. The result of matchLabels and matchExpressions are ANDed. An empty label selector matches all objects. A null label selector matches no objects.

It sounds to me like this is really another instance where #77 comes to bear.

**Comment by arianvp** [comment]
I think this is an instantiation of https://stackoverflow.com/questions/50309057/what-is-the-purpose-of-a-kubernetes-deployment-pod-selector (See first answer).  It used to be the case that `selector` would infer a value that matches all the `labels` on the `Deployment` but that behavior was removed and now the user is expected to create a `selector` themselves.  Because by default this would mean _all_ labels were part of the selector, which made changing labels after the fact difficult.

About it being an instantiaton of #77 . This is a bit more funky than that.  In this case it is valid for both of them to be defined or one of them to be defined.  #77 is about mutually exclusive values mostly.

The `anyOf` operator in `OpenAPI v3` would describe these semantics of it nicely. https://swagger.io/docs/specification/data-models/oneof-anyof-allof-not/#anyof

However I'm not sure how we would translate an `anyOf` to a dhall type.  `oneOf` is easy as it maps nicely to a sum type, but for `anyOf` we need something more fancy.

**Comment by arianvp** [comment]
The other ugly thing here is that `matchExpression` is stricly more general than `matchLabels`.  That is, you can do everything with the `matchExpresison` field that you can do with the `matchLabels` field.   Ideally, `matchLables` wouldn't exist at all, and would be a _function_ of the type:

```haskell
matchLabels  :: [(string, string)] ->  [LabelSelectorRequirement]
```

e.g.:
```yaml
matchLabels:
  app: arians-cool-app
  env: production
```
is the same as:

```yaml
matchExpressions:
  - key: app
    operator: In
    values: 
    - arians-cool-app
  - key: env
    operator: In
    values:
      - production
```

**Comment by Gabriella439** [comment]
This is fixed by #110 as the `selector` field is now required
